demos/03 Image Segmentation/03b Training a Segmentation Model.ipynb

```diff
@@ -0,0 +1,444 @@
+    "train_masks = mask_array[:800]\n",
+    "val_masks = mask_array[800:]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f38c5f46-0b5a-49f2-857e-43eb9c7e41b5",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "train_data.shape, val_data.shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f68c2ab2-f638-416c-9fd4-7e2e39fed1ed",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "train_masks.shape, val_masks.shape"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9b25c54d-c156-4304-8cda-7fe054c0f198",
+   "metadata": {},
+   "source": [
+    "## Training a Computer Vision Model\n",
+    "A sample U-Net architecture, commonly used in image segmentation tasks. In practice this is where you should input your own methodology and iterate!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d334c7d1-47fd-459c-96a5-d010c3fd6e79",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "inputs = Input(shape=(None, None, 3))\n",
+    "\n",
+    "conv1 = Conv2D(64, 3, activation=\"relu\", padding=\"same\")(inputs)\n",
+    "pool1 = MaxPooling2D(pool_size=(2, 2))(conv1)\n",
+    "conv2 = Conv2D(128, 3, activation=\"relu\", padding=\"same\")(pool1)\n",
+    "pool2 = MaxPooling2D(pool_size=(2, 2))(conv2)\n",
+    "conv3 = Conv2D(256, 3, activation=\"relu\", padding=\"same\")(pool2)\n",
+    "pool3 = MaxPooling2D(pool_size=(2, 2))(conv3)\n",
+    "conv4 = Conv2D(512, 3, activation=\"relu\", padding=\"same\")(pool3)\n",
+    "pool4 = MaxPooling2D(pool_size=(2, 2))(conv4)\n",
+    "\n",
+    "conv5 = Conv2D(1024, 3, activation=\"relu\", padding=\"same\")(pool4)\n",
+    "\n",
+    "up6 = Concatenate()([UpSampling2D(size=(2, 2))(conv5), conv4])\n",
+    "conv6 = Conv2D(512, 3, activation=\"relu\", padding=\"same\")(up6)\n",
+    "up7 = Concatenate()([UpSampling2D(size=(2, 2))(conv6), conv3])\n",
+    "conv7 = Conv2D(256, 3, activation=\"relu\", padding=\"same\")(up7)\n",
+    "up8 = Concatenate()([UpSampling2D(size=(2, 2))(conv7), conv2])\n",
+    "conv8 = Conv2D(128, 3, activation=\"relu\", padding=\"same\")(up8)\n",
+    "up9 = Concatenate()([UpSampling2D(size=(2, 2))(conv8), conv1])\n",
+    "conv9 = Conv2D(64, 3, activation=\"relu\", padding=\"same\")(up9)\n",
+    "\n",
+    "outputs = Conv2D(1, 1, activation=\"sigmoid\")(conv9)\n",
+    "\n",
+    "model = Model(inputs=inputs, outputs=outputs)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "062922c6-2ae0-400d-960a-993a9e402f5d",
+   "metadata": {},
+   "source": [
+    "Compile:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5e2afbe0-e571-47ef-832d-1037fcaf5518",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "model.compile(optimizer=\"adam\", loss=\"binary_crossentropy\", metrics=[\"accuracy\"])"
+   ]
+  },
+  {
```

Comment:
```
class CustomUNet:
    def __init__(self, optimizer="adam", activation="sigmoid", loss="binary_crossentropy"):
        self.optimizer = optimizer
        self.activation = activation
        self.loss = loss
        self.model = self.build_model()

    def build_model(self):
        # Define input layer
        inputs = Input(shape=(None, None, 3))
        
        # Encoder
        conv1 = Conv2D(64, 3, activation="relu", padding="same")(inputs)
        pool1 = MaxPooling2D(pool_size=(2, 2))(conv1)
        conv2 = Conv2D(128, 3, activation="relu", padding="same")(pool1)
        pool2 = MaxPooling2D(pool_size=(2, 2))(conv2)
        conv3 = Conv2D(256, 3, activation="relu", padding="same")(pool2)
        pool3 = MaxPooling2D(pool_size=(2, 2))(conv3)
        conv4 = Conv2D(512, 3, activation="relu", padding="same")(pool3)
        pool4 = MaxPooling2D(pool_size=(2, 2))(conv4)

        # Center
        conv5 = Conv2D(1024, 3, activation="relu", padding="same")(pool4)

        # Decoder
        up6 = Concatenate()([UpSampling2D(size=(2, 2))(conv5), conv4])
        conv6 = Conv2D(512, 3, activation="relu", padding="same")(up6)
        up7 = Concatenate()([UpSampling2D(size=(2, 2))(conv6), conv3])
        conv7 = Conv2D(256, 3, activation="relu", padding="same")(up7)
        up8 = Concatenate()([UpSampling2D(size=(2, 2))(conv7), conv2])
        conv8 = Conv2D(128, 3, activation="relu", padding="same")(up8)
        up9 = Concatenate()([UpSampling2D(size=(2, 2))(conv8), conv1])
        conv9 = Conv2D(64, 3, activation="relu", padding="same")(up9)

        # Output layer
        outputs = Conv2D(1, 1, activation=self.activation)(conv9)

        # Define the model
        model = Model(inputs=inputs, outputs=outputs)

        # Compile the model
        model.compile(optimizer=self.optimizer, loss=self.loss, metrics=["accuracy"])

        return model

# Example usage:
unet_model = CustomUNet()
model = unet_model.model
```

If you wanted to have this as a class, you can replace these cells with this code block. Its not necessary though!

Reply:
works great and looks much nicer, thank you!